File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.cpp
#include <iostream>
using namespace std;

// num = 상자번호, w = 한 층 최대 상자 수, n = 전체 상자 수
int solution(int n, int w, int num) {
    int answer = 0;

    int floor = (num + w - 1) / w;       // 현재 층 (1부터)
    int locate = (num - 1) % w;          // 해당 층 내 위치
    // 홀수층 짝수층 번호 맞추기
    int col = (floor % 2 == 1) ? locate : (w - 1 - locate);

    int maxfloor = (n + w - 1) / w;      // 총 층 수

    for (int i = floor; i <= maxfloor; ++i) {
        int BoxfloorCount; // 마지막 층 상자 개수
        if (i == maxfloor) {
            BoxfloorCount = n - (i - 1) * w;
        } else {
            BoxfloorCount = w;
        }

        if (i % 2 == 1) {
            if (col <= BoxfloorCount) answer++;
        } else {
            if (col >= w - BoxfloorCount) answer++;
        }
    }
    return answer;
}

File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.py
def solution(n, w, num):
    answer = 0
    arr = []
    for i in range(1, (n//w)*w+1, w):
        if (i // w)%2 == 0:
            arr.append([j for j in range(i, i+w)])
        else:
            arr.append([j for j in range(i+w-1,i-1,-1)])
    if n%w != 0:
        if (n//w)%2 == 0:
            arr.append([j for j in range((n//w)*w, n+1)])
            arr[n//w].extend([0 for j in range(w-n%w)])
        else:
            arr.append([0 for j in range(w-n%w)])
            arr[n//w].extend([j for j in range(n, (n//w)*w, -1)])

    found = False
    for aa in reversed(arr):
        answer += 1
        for i, a in enumerate(aa):                
            if a == num:
                found = True
                break
        if found:
            if arr[-1][i] == 0:
                answer -= 1
            break

    if found == False:
        answer = 0

    return answer